### ISSUE TITLE
Forcefully deactivate grains stuck in a request for too long

### ISSUE THREAD

**Issue by sergeybykov** [dotnet/orleans#1383]
We've seem this happening multiple times - a grain call gets stuck forever either awaiting a `Task` that never completes or, very rarely, trying to retrieve `.Result` from a resolved `Task`.

The general problem here is that, whatever the reason for a grain to get stuck in a method call, it is forever unavailable for other calls, until the silo gets restarted. Hence the user, session, device the grain represents gets unavailable as the result. We've seen occasionally a grain being unavailable for a couple of days.

The proposal is to add logic for forceful deactivation of such grains either automatically or/and via an explicit command. This would allow a new activation of the stuck grain to be created to process future requests.

The automatic option could be such that we set a configurable threshold for request execution time, e.g. 10 minutes, and automatically deactivate grains that reach it. The explicit command option can be implemented via ManagementGrain

The potential risk of forceful deactivation with incomplete execution of request is that the grain may eventually resume execution of the method if the `Task` or a deadlock it got blocked on gets resolved. This may lead to a violation of the single-threading/single-activation guarantees.

Weighing the risk against the experience of grains getting unavailable in most such cases until the silo is rebooted, I think the pros outweigh the cons here.

**Comment by ReubenBond** [comment]
Given Orleans controls the task scheduling, couldn't we ensure that the continuation is never scheduled in the case that the grain has been forcefully deactivated?

**Comment by gabikliot** [comment]
I think it is important issue and we can definitely do a lot in this direction (and already have a lot of pieces available). Before I provide more details, just wanted to clarify:

>a grain call gets stuck forever awaiting a `Task` that never completes

Here you mean not a grain call, but some arbitrary library call, which does not have an enforced timeout?

> a grain call gets stuck forever ... to retrieve .Result from a resolved Task.

How is that even possible? How a call to `.Result` from a resolved `Task` can block forever?

@ReubenBond , yes of course we can do that, and already do exactly that for deactivated grains that by mistake left not-awaited ContinueWith blocks. What @sergeybykov means, beyond what you wrote, is the case when the grain code is stuck in the middle of some code, not awaiting a task but executing (or blocked on some synchronous system call for example), so you either have to preempt a thread, or deal with the fact that after the code unblocks, it will keep executing and we can't stop it, until it calls await again, at which point we can stop it.

**Comment by sergeybykov** [comment]
>Here you mean not a grain call, but some arbitrary library call, which does not have an enforced timeout?

Yes, not a grain call but a call to library that returns a `Task` that never gets resolved.

>How is that even possible? How a call to  .Result  from a resolved  Task  can block forever?

I suspect it's a bug somewhere in TPL. We saw this happening periodically, very infrequently, in one of 343 services where the code would do `Task.WhenAll(tasks)` and then run a `foreach` loop over `tasks` retrieving values of `Results` properties of the tasks. Even though the tasks were expected to be resolved, it would sometimes stuck in the `.Result` call for a couple of days until they rebooted the silo.

We are investigating an issue in another 343 services that looks very similar.

**Comment by gabikliot** [comment]
Here is a list of what we already have and what we can do more.

First, a **definition**:
**Routable Activation**: an activation A of grain G is called routable if A appears as an activation for G in the grain directory or at least in one of the grain directory caches on one of the silos.

**Invariant**: any new request that enters the system can never be routed and executed on a non-routable activation. That holds also for an activation that may still be alive and active on some silo. So if lets say an activation is somehow stuck on some silo in a VALID activation state, but is not-routable, no new request will arrive to it, and any new request to grain G will arrive to (potentially activating) a new activation.

**We support this invariant already now**. That is what helps us to be very clear about what should "eventually be correct" - only the directory and the directory caches, since only those are used to finding an activation. If you find out that this invariant is violated anywhere, it is a bug and should be fixed. 

So given the above invariant what we can do:

A - what we have now, do nothing more:
1) OrleansManager already now supports deleting an activation from a directory. So one can just call it, and then wait until all caches expire (I think the default is only 4 minutes, configurable) and after that the stuck activation becomes non-routable and thus any new request will arrive to a new activation.
2) If you do nothing on top on this, you get a new activation after that time, but the old might still be stuck either blocking a valuable thread, or maybe resuming sometime later and finishing the current request. Also, all the request that were queued on it until be became non-routable - most probably they will already expire and not execute, but if not, they will execute on the old activation, since we don't check again on every request before dispatching it if the activation is still routable. We could.
3) OrleansManager also has a call grainReport, that outputs all we know about that grain - all activation, all caches, directory. So you can use that to validate that indeed directory was unregistered correctly, cache expired, ... One can even write automated tests with a stuck grain and uses the calls from OrleansManager  and validate that indeed the stuck activation is non routble, and new requests arrive to a new activation.

B -  in addition to A also shrink the cache expiration time:
4) If you want, you can shrink the cache expiration time by firing parallel calls to all silos to clear their directory cache for that grain.  That will require a support in OrleansManager  and SiloControl, but is very easy and has very low risk, since clearing the cache should always be safe.

C - in addition to A and B also speed up deactivation of a stuck activation:
4) After unregistered from the directory, we can also send a Deactivate Request to the Catalog to deactivate the stuck activation. That becomes more complicated and will involve some changes in the activation lifecycle. Now when we call `DeactivateActivationOnIdle` we wait until the current request is done. But that is exactly what you don't want to do in this case. What we do now depends on whether the activation is actually executing on some thread or awating.
4.a) Lets look at awaiting first. You can know that based on the state of the activation (work item group) in the scheduler. Now you will basically need to set the work item group to not execute any more Tasks on this activation, and proceed with the regular deactivation sequence. Probably also don't want to call application's `OnDeactivateAsync` since that would be semantic violation - just abruptly stopping is like a crash, but abruptly stopping plus `OnDeactivateAsync` is not possible.
4.b) if you also want to handle the stuck executing code case (the most delicate case), here I would interrupt the worker thread that executes this activation, join the thread (or any other means to make sure this thread is done with this activation) and then proceed with 4.a. If this thread is dead now, inject a new thread ONLY AFTER you are sure this thread is dead/aborted/finished. Otherwise, you may end up injecting new threads while other threads are still executing, and this may create too many threads, that will end up doing more damage than good.

**Comment by gabikliot** [comment]
Any feedback @sergeybykov ? Was that useful/enough details?

**Comment by sergeybykov** [comment]
@gabikliot Thanks for such a detailed write-up!

I'm looking at this priority wise almost bottom up. D was the motivation for opening this issue - automatically detect cases of stuck grains and take measures before the unhappy users whos grains became unavailable hours or days go complain to support.

If we detect and simply remove a stuck activation from the directory (A), automating what today is the manual process with OrleansManager.exe, that will already be a huge win because the users won't experience outage or degradation of their experience for much longer than the detection window. I'm not sure where the default threshold should be: 10-15-20-30 minutes, whatever we choose there will be better than users being locked out for hours and days.

If at the same time we purge caches (B), that may be a nice and cheap optimization.

C is pri 2 to me. Nice to have, but not required. So long as the activation isn't routable, and it's a very rare case (as we see today - a handful grains over the course of days and weeks), I don't see much harm in leaving it alone. If grains get stuck much more frequently, the app code needs to be fixed anyway.

**Comment by gabikliot** [comment]
You are welcome @sergeybykov .

I agree that "from the business side" (feature importance) you would want to start with D and see C as an optimization. But from correctness wise (safety), if we start with D than C becomes non optimization priority one safety measure.
The problem is you can never be 100% sure your detection of a stuck activation in D is 100% correct. You will have miss-detection. GC will pause for long time, or you will have a bug in detection of idleness, what ever. Or it was detected as stuck but was unstuck the next moment you detected it as stuck. Or, if you are not convinced by my hypothetical examples above, imagine the call is stuck on this non-re-entrant grain, but the time is still ticking.

The bottom line is - lets just assume that for now - the detection will not be perfect.
So you detected and if now the next thing you do is A - unregister in the directory but without C (that is, without deactivating the stuck activation first), you are now at a risk of 2 activations. And this is now not a failure silo case.

So, my point is, if you want to do automatic detection and resolution, you need to:
- detect (D)
- deactivate locally first, stop the activation fully, if it is stuck on the thread - interrupt this thread and join it until it aborts (C)
- only now unregister from the directory (A). No need for B now.

If on the other hand you stay with the manual solution, then the operator knows that he messes up with the system, he is ready mentally to have 2 activations, he is watching it. Now C becomes an optimization.

**Comment by sergeybykov** [comment]
Resolved via #2387.
